api/client/build.go

```diff
@@ -315,3 +277,295 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+// path of the dockerfile in that context directory, and a non-nil error on
+// success.
+func getContextFromURL(out io.Writer, remoteURL, dockerfileName string) (absContextDir, relDockerfile string, err error) {
+	response, err := httputils.Download(remoteURL)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to download remote context %s: %v", remoteURL, err)
+	}
+	defer response.Body.Close()
+
+	// Pass the response body through a progress reader.
+	progReader := &progressreader.Config{
+		In:        response.Body,
+		Out:       out,
+		Formatter: streamformatter.NewStreamFormatter(),
+		Size:      int(response.ContentLength),
+		NewLines:  true,
+		ID:        "",
+		Action:    fmt.Sprintf("Downloading build context from remote url: %s", remoteURL),
+	}
+
+	return getContextFromReader(progReader, dockerfileName)
+}
+
+// getContextFromLocalDir uses the given local directory as context for a
+// `docker build`. Returns the absolute path to the local context directory,
+// the relative path of the dockerfile in that context directory, and a non-nil
+// error on success.
+func getContextFromLocalDir(localDir, dockerfileName string) (absContextDir, relDockerfile string, err error) {
+	// When using a local context directory, when the Dockerfile is specified
+	// with the `-f/--file` option then it is considered relative to the
+	// current directory and not the context directory.
+	if dockerfileName != "" {
+		if dockerfileName, err = filepath.Abs(dockerfileName); err != nil {
+			return "", "", fmt.Errorf("unable to get absolute path to Dockerfile: %v", err)
+		}
+	}
+
+	return getDockerfileRelPath(localDir, dockerfileName)
+}
+
+var dockerfileFromLinePattern = regexp.MustCompile(`(?i)^FROM[ \f\r\t\v]+(?P<image>[^ \f\r\t\v\n#]+)`)
+
+type trustedDockerfile struct {
+	*os.File
+	size int64
+}
+
+func (td *trustedDockerfile) Close() error {
+	td.File.Close()
+	return os.Remove(td.File.Name())
+}
+
+// rewriteDockerfileFrom rewrites the given Dockerfile by resolving images in
+// "FROM <image>" instructions to a digest reference. `translator` is a
+// function that takes a repository name and tag reference and returns a
+// trusted digest reference.
+func rewriteDockerfileFrom(dockerfileName string, translator func(string, registry.Reference) (registry.Reference, error)) (newDockerfile *trustedDockerfile, err error) {
+	dockerfile, err := os.Open(dockerfileName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open Dockerfile: %v", err)
+	}
+	defer dockerfile.Close()
+
+	scanner := bufio.NewScanner(dockerfile)
+
+	// Make a tempfile to store the rewritten Dockerfile.
+	tempFile, err := ioutil.TempFile("", "trusted-dockerfile-")
+	if err != nil {
+		return nil, fmt.Errorf("unable to make temporary trusted Dockerfile: %v", err)
+	}
+
+	defer func() {
+		if err != nil {
+			// Close the tempfile if there was an error during Notary lookups.
+			// Otherwise the caller should close it.
+			tempFile.Close()
+			os.Remove(tempFile.Name())
```

Comment:
We want to make sure that the file is removed during this function *only* if there's an error (we don't want to close it before the caller has a chance to use it). When the caller calls the `Close()` method (defined on the custom struct type above) it will then close and remove this temp file.